releases/release-4.0.12.md

```diff
@@ -0,0 +1,117 @@
+    - Use an adaptive aggressive GC strategy to reduce memory usage
+
+## Bug Fixes
+
++ TiDB
+
+    - Fix get var expr when session var is hex literal [#23372](https://github.com/pingcap/tidb/pull/23372)
+    - Fix the bug that wrong collation is used when try fast path for enum or set [#23292](https://github.com/pingcap/tidb/pull/23292)
+    - Wrong result of nullif expr when used with is null expr [#23279](https://github.com/pingcap/tidb/pull/23279)
+    - Statistics: fix a case that auto-analyze is triggered outside its time range [#23219](https://github.com/pingcap/tidb/pull/23219)
+    - Executor: fix cast function will ignore the error for point get key construction [#23211](https://github.com/pingcap/tidb/pull/23211)
+    - Fixed a bug that prevented SPM from taking effect [#23209](https://github.com/pingcap/tidb/pull/23209)
+    - Fix wrong table filters for index merge plan [#23165](https://github.com/pingcap/tidb/pull/23165)
+    - Fix unexpected NotNullFlag in case when expr ret type [#23135](https://github.com/pingcap/tidb/pull/23135)
+    - Fix a bug that collation is not handle for text type [#23092](https://github.com/pingcap/tidb/pull/23092)
+    - Fix range partition prune bug for IN expr [#23074](https://github.com/pingcap/tidb/pull/23074)
+    - Fix request block forever when replace a tombstone store with the store has same IP addr [#23071](https://github.com/pingcap/tidb/pull/23071)
+    - Do not ajust int when it is null and compared year [#22844](https://github.com/pingcap/tidb/pull/22844)
+    - Fix load data lost connection error on tables with auto_random column [#22736](https://github.com/pingcap/tidb/pull/22736)
+    - Fix ddl hang over when it meets panic in cancelling path. [#23297](https://github.com/pingcap/tidb/pull/23297)
+    - Fix wrong key range of index scan when filter is comparing year column with NULL [#23104](https://github.com/pingcap/tidb/pull/23104)
+    - Fix create view success but failed when using it [#23083](https://github.com/pingcap/tidb/pull/23083)
+
++ PD
+
+    - Save to the region cache when pending-peers or down-peers change [#3471](https://github.com/pingcap/pd/pull/3471)
+    - Checker: prevent the regions in split-cache from becoming the target of merge [#3459](https://github.com/pingcap/pd/pull/3459)
+
++ TiFlash
+
+    - Fix an issue of incorrect results when the default value of `binary` type column contains leading or tailing zero bytes
+    - Fix the bug that TiFlash fail to sync schema if the name of database contains special characters
+    - Fix an issue of incorrect results when handling `IN` expression with decimal values
+    - Fix the bug that metrics about opened file count shown in Grafana is high
+    - Fix the bug that TiFlash does not support `Timestamp` literal
+    - Fix the potential crash while handling `FROM_UNIXTIME` expression
+    - Fix an issue of incorrect results when casting string as integer
+    - Fix a bug that `like` function may return wrong result
+
```

Comment:
+ TiKV
    - Fix the issue that the IN expr(coprocessor) didn't handle unsigned/signed int properly [#9850](https://github.com/tikv/tikv/pull/9850)
    - Fix the issue that the ingests operation cannot reentrant. [#9779](https://github.com/tikv/tikv/pull/9779)
    - Fix the issue that the space missed when converting json to string in TiKV coprocessor [#9666](https://github.com/tikv/tikv/pull/9666)